include/picongpu/plugins/RandomizedParticleMerger/RandomizedParticleMerger.kernel

```diff
@@ -0,0 +1,513 @@
+
+                    break;
+                }
+            }
+        }
+
+
+        /** Entry point of the particle merging algorithm
+         *
+         * @param cellIndex n-dim. cell index from the origin of the local domain
+         */
+        template< typename T_Acc>
+        DINLINE void operator()(
+            T_Acc const & acc,
+            const pmacc::math::Int<simDim>& cellIndex
+        )
+        {
+            /* multi-dim vector from origin of the super cell to a cell in units of cells */
+            const pmacc::math::Int<simDim> threadIndex = cellIndex % SuperCellSize::toRT();
+
+            /* conversion from a multi-dim cell coordinate to a linear coordinate
+             * of the cell in its super cell */
+            const int linearThreadIdx = pmacc::math::linearize(
+                pmacc::math::CT::shrinkTo<SuperCellSize, simDim-1>::type::toRT(),
+                threadIndex
+            );
+            
+          
+            /* fixed-sized array of Voronoi cells */
+            PMACC_SMEM( acc, listVoronoiCells, ArrayVoronoiCells );
+            /* holds indices of active Voronoi cells within `listVoronoiCells` */
+            PMACC_SMEM( acc, voronoiIndexPool, VoronoiIndexPool );
+
+            /* number of initial Voronoi cells
+             *
+             * `1u << simDim` is equivalent to `pow(2, simDim)` but can be
+             * calculated at compile-time to save a shared variable.
+             */
+            constexpr uint16_t numInitialVoronoiCells = pmacc::math::CT::volume<
+                SuperCellSize
+            >::type::value / ( 1u << simDim );
+            
+            pmacc::math::Int<simDim> localOffset = cellIndex / SuperCellSize::toRT() - guardSuperCells; 
+            constexpr uint32_t numWorkers = pmacc::math::CT::volume< SuperCellSize >::type::value;
+            pmacc::mappings::threads::WorkerCfg<numWorkers> workerCfg( linearThreadIdx );
+            
+            auto generator = rngFactory( acc, localOffset, workerCfg );
+
+            if( linearThreadIdx == 0 )
+            {
+                /* init index pool of Voronoi Cells */
+                voronoiIndexPool = VoronoiIndexPool( numInitialVoronoiCells );
+            }
+
+            __syncthreads();
+
+            /* set initial Voronoi cells into `collecting` state */
+            if( linearThreadIdx < numInitialVoronoiCells )
+                listVoronoiCells[linearThreadIdx] = VoronoiCell();
+
+            /* init the voronoiCellId attribute for each particle */
+            this->initVoronoiCellIdAttribute( acc, cellIndex );
+			
+            /* main loop of the merging algorithm */
+            while( voronoiIndexPool.size() > 0 )
+            {         
+                this->processParticles(
+                    acc,
+                    cellIndex,
+                    listVoronoiCells
+                );
+
+                __syncthreads();
+
+                /* TODO: parallelize */
```

Comment:
please use `@todo` that doxygen can generate a todo list

Reply:
Agree, tho this one was copied from the original implementation (however should still be changed here).